x-pack/solutions/security/plugins/security_solution/server/lib/entity_analytics/entity_store/tasks/snapshot/snapshot.ts

```diff
@@ -0,0 +1,312 @@
+  } catch (e) {
+    logger.warn(`[Entity Store]  [task ${taskId}]: error scheduling task, received ${e.message}`);
+    throw e;
+  }
+}
+
+export async function removeEntityStoreSnapshotTask({
+  logger,
+  namespace,
+  entityType,
+  taskManager,
+}: {
+  logger: Logger;
+  namespace: string;
+  entityType: EntityType;
+  taskManager: TaskManagerStartContract;
+}) {
+  try {
+    await taskManager.remove(getTaskId(namespace, entityType));
+    logger.info(`[Entity Store]  Removed snapshot task for ${entityType} in ${namespace}`);
+  } catch (err) {
+    if (!SavedObjectsErrorHelpers.isNotFoundError(err)) {
+      logger.error(`[Entity Store]  Failed to remove snapshot task: ${err.message}`);
+      throw err;
+    }
+  }
+}
+
+export async function runTask({
+  logger,
+  telemetry,
+  context,
+  esClientGetter,
+}: {
+  logger: Logger;
+  telemetry: AnalyticsServiceSetup;
+  context: RunContext;
+  esClientGetter: () => Promise<ElasticsearchClient>;
+}): Promise<{
+  state: EntityStoreFieldRetentionTaskState;
+}> {
+  const state = context.taskInstance.state as EntityStoreFieldRetentionTaskState;
+  const taskId = context.taskInstance.id;
+  const log = entityStoreTaskLogFactory(logger, taskId);
+  const debugLog = entityStoreTaskDebugLogFactory(logger, taskId);
+  const esClient = await esClientGetter();
+  try {
+    const taskStartTime = moment().utc();
+    const snapshotDate = rewindToYesterday(taskStartTime.toDate());
+    log('running task');
+
+    const entityType = context.taskInstance.params.entityType as EntityType;
+    const namespace = context.taskInstance.params.namespace as string;
+    if (namespace === '') {
+      const err = `Task ${taskId} expected vaild namespace in params, got ""`;
+      log(err);
+      throw err;
+    }
+    const updatedState = {
+      lastExecutionTimestamp: taskStartTime.toISOString(),
+      lastSnapshotTookSeconds: 0,
+      namespace,
+      entityType: entityType as string,
+      runs: state.runs + 1,
+    };
+
+    if (taskId !== getTaskId(namespace, entityType)) {
+      log('outdated task; exiting');
+      return { state: updatedState };
+    }
+
+    debugLog('creating snapshot index');
+    const { index: snapshotIndex } = await createEntitySnapshotIndex({
+      esClient,
+      entityType,
+      namespace,
+      snapshotDate,
+    });
+
+    debugLog(`reindexing entities to ${snapshotIndex}`);
+    await esClient.reindex({
+      source: {
+        index: [getEntitiesIndexName(entityType, namespace)],
+      },
+      dest: {
+        index: snapshotIndex,
+      },
+      conflicts: 'proceed',
+    });
+
+    const resetIndex = getEntitiesResetIndexName(entityType, namespace);
+    debugLog(`reindexing entities to ${resetIndex}`);
+    await esClient.reindex({
+      source: {
+        index: [getEntitiesIndexName(entityType, namespace)],
+      },
+      dest: {
+        index: resetIndex,
+      },
+      conflicts: 'proceed',
+      script: {
```

Comment:
I understand we are publishing a reset with date only fields right now.  This isn't very clear at the first read though.

it's a bit complex to read that we get all entities from index, remove most data, keep a few, because that will be picked up by entity store transform in the next run.

We need to find away to tell this story better in code

Reply:
Are the comments not enough? If so, I can add one that explicitly states so.

Reply:
[Comment added](https://github.com/elastic/kibana/pull/233970/files#diff-07891634b1284d348e0eb01f5bde39fcf0e4c2602c76a74abf8484541ba62b22R137-R139)

Reply:
We could consider putting Kubas diagram from the PR description in a README somewhere, I found it very useful for understanding the feature